cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1152,45 +1158,93 @@ void writer::impl::write(table_view const& table)
   // compression/decompression performance).
   using cudf::io::parquet::gpu::max_page_fragment_size;
 
-  size_type const num_fragments = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
+  bool row_group_sizes_specified = !row_group_sizes.empty();
+  size_type num_fragments        = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
+  if (row_group_sizes_specified) {
+    num_fragments = 0;
+    for (std::size_t i = 0; i < row_group_sizes.size(); i++) {
+      num_fragments += (row_group_sizes[i] + max_page_fragment_size - 1) / max_page_fragment_size;
+    }
+  }
   cudf::detail::hostdevice_2dvector<gpu::PageFragment> fragments(
     num_columns, num_fragments, stream);
 
+  if (row_group_sizes_specified) {
+    // auto fragments_span = host_2dspan<gpu::PageFragment>{fragments};
+    size_type i = 0;
```

Comment:
Can we use more meaningful names for the indices here? It doesn't appear that this function is using a traditional 3d indexing scheme like `arr[i][j][k]`, and I found the uses of `i`, `j`, `k` somewhat hard to follow without knowing what they represent.